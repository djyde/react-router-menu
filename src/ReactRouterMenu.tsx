import * as React from 'react'
import {
  Menu
} from 'antd'
import { PlainRoute, Link } from 'react-router'
import { MenuProps } from 'antd/lib/menu'
import { LocationState } from "history"

const MenuItem = Menu.Item
const SubMenu = Menu.SubMenu

export type RouteCallback = (err: any, route: PlainRoute) => void
export type RoutesCallback = (err: any, routesArray: PlainRoute[]) => void

export interface IReactRouterMenuRoute extends PlainRoute {
  path: string,
  title: string,
  childRoutes?: IReactRouterMenuRoute[],
  getChildRoutes?(partialNextState: LocationState, callback: RoutesCallback): void,
  indexRoute?: IReactRouterMenuRoute,
  getIndexRoute?(partialNextState: LocationState, callback: RouteCallback): void
}

const createMenuItem = (route: IReactRouterMenuRoute) => {
  if (route.getChildRoutes || route.getIndexRoute) {
    console.warn('getChildRoutes and getIndexRoute are not supported yet.')
    return
  }
  if (route.path === '/') {
    return createMenuItemFromRoutes(route.childRoutes)
  }
  if (!route.childRoutes && !route.getChildRoutes) {
    return (
      <MenuItem key={route.path}>
        <Link to={route.path}>
          {route.title}
        </Link>
      </MenuItem>
    )
  } else {
    return (
      <SubMenu key={route.path} title={route.title}>
        {createMenuItemFromRoutes(route.childRoutes)}
      </SubMenu>
    )
  }
}

const createMenuItemFromRoutes = (childRoutes: IReactRouterMenuRoute[] = []) => {
  return childRoutes.map(route => {
    return createMenuItem(route)
  })
}

export const createReactRouterMenu = (route: IReactRouterMenuRoute) => {
  const r = createMenuItem(route)
  return (props: MenuProps) => (
    <Menu {...props}>
      {r}
    </Menu>
  )
}
