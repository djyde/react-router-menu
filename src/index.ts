export { createReactRouterMenu, IReactRouterMenuRoute } from './ReactRouterMenu'
