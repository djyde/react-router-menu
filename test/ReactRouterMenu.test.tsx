import * as React from 'react'
import { createReactRouterMenu } from '../src'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { Router, hashHistory } from 'react-router'

describe('createReactRouterMenu', () => {

  const routes = {
    title: 'App',
    path: '/',
    component: App,
    childRoutes: [
      { title: 'Home', path: 'home', component: () => <div>Home</div> },
      { title: 'About', path: 'about', component: () => <div>About</div> }
    ]
  }

  function App({ children }) {
    const ReactRouterMenu = createReactRouterMenu(routes)
    return (
      <div>
        <ReactRouterMenu mode='horizontal' />
        <div>
          {children}
        </div>
      </div>
    )
  }

  const Root = <Router history={hashHistory} routes={routes} />

  it('snapshot', () => {
    let tree = renderer.create(Root)
    tree.toJSON()
    expect(tree).toMatchSnapshot()
  })
})