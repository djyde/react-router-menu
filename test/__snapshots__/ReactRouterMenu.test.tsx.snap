// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReactRouterMenu snapshot 1`] = `
<div>
  <ul
    aria-activedescendant=""
    className="ant-menu ant-menu-horizontal ant-menu-light ant-menu-root"
    onKeyDown={[Function]}
    role="menu"
    style={Object {}}
    tabIndex="0"
  >
    <li
      className="ant-menu-submenu-horizontal ant-menu-submenu"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        aria-expanded={false}
        aria-haspopup="true"
        aria-owns="/$Menu"
        className="ant-menu-submenu-title"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={Object {}}
      >
        App
      </div>
    </li>
  </ul>
  <div />
</div>
`;
